fsgrid.hpp

```diff
@@ -26,99 +26,259 @@
-         int n_per_task = globalCells/ntasks;
-         int remainder = globalCells%ntasks;
-         if(my_n < remainder) {
-            return n_per_task+1;
+   //! Helper function: calculate size of the local coordinate space for the given dimension
+   // \param globalCells Number of cells in the global Simulation, in this dimension
+   // \param ntasks Total number of tasks in this dimension
+   // \param my_n This task's position in this dimension
+   // \return Number of cells for this task's local domain (actual cells, not counting ghost cells)
+   static FsIndex_t calcLocalSize(FsSize_t globalCells, Task_t ntasks, Task_t my_n) {
+      FsIndex_t n_per_task = globalCells/ntasks;
+      FsIndex_t remainder = globalCells%ntasks;
+      if(my_n < remainder) {
+         return n_per_task+1;
+      } else {
+         return n_per_task;
+      }
+   }
+
+   //! Helper function to calculate decomposition of files without stored decompositions
+   static void computeLegacyDomainDecomposition(const std::array<FsSize_t, 3>& GlobalSize, Task_t nProcs, std::array<Task_t,3>& processDomainDecomposition, int stencilSize=1, int verbose = 0) {
+      int myRank, MPI_flag;
+      MPI_Initialized(&MPI_flag);
+      if(MPI_flag){
+         MPI_Comm_rank(MPI_COMM_WORLD, &myRank); // TODO allow for separate communicator
+      } else {
+         myRank = 0;
+      }
+
+      std::array<FsSize_t, 3> systemDim;
+      std::array<FsSize_t, 3> processBox;
+      std::array<FsSize_t, 3> minDomainSize;
+      int64_t optimValue = std::numeric_limits<int64_t>::max();
+      std::vector<std::pair<int64_t,std::array<Task_t,3>>> scored_decompositions;
+      scored_decompositions.push_back(std::pair<int64_t,std::array<Task_t,3>>(optimValue,{0,0,0}));
+      for(int i = 0; i < 3; i++) {
+         systemDim[i] = GlobalSize[i];
+         if(GlobalSize[i] == 1) {
+            // In 2D simulation domains, the "thin" dimension can be a single cell thick.
+            minDomainSize[i] = 1;
          } else {
-            return n_per_task;
+            // Otherwise, it needs to be at least as large as our ghost
+            // stencil, so that ghost communication remains consistent.
+            minDomainSize[i] = stencilSize;
+         }
+      }
+      processDomainDecomposition = {1, 1, 1};
+      for (Task_t i = 1; i <= std::min(nProcs, (Task_t)(GlobalSize[0]/minDomainSize[0])); i++) {
+         for (Task_t j = 1; j <= std::min(nProcs, (Task_t)(GlobalSize[1]/minDomainSize[1])) ; j++) {
+            if( i * j  > nProcs ){
+               break;
+            }
+            Task_t k = nProcs/(i*j);
+            // No need to optimize an incompatible DD, also checks for missing remainders
+            if( i * j * k != nProcs ) 
+               continue;
+
+            processBox[0] = systemDim[0]/i;
+            processBox[1] = systemDim[1]/j;
+            processBox[2] = systemDim[2]/k;
+
+            int64_t value = 
```

Comment:
I've been thinking about this further. I fear that for the legacy mode, we need to go back to floats, as much as I hate their use here.

The reason is that if the FSgrid domain does not decompose exactly into the number of tasks in that dimension (and thus some tasks have a remainder), the `value`used to be calculated with the non-integer value resulting of this decomposition. Thus, by enforcing integers, we might not end up with the old legacy value after all.

Reply:
I believe this is what I initially suggested ;).

Or do we just scrap this entirely and finagle a way to read the FsGrid metadata from the one the VisIt plugin is reading?

Reply:
Perhaps, but now I believe the reason it needs to be has been clearly stated. And then indeed that visit plugin data comes along to rejig everything!

If we in the future intend to drop the full mesh data, then we will still have the decomposition to read, so that's safe. And any data generated with the current version will have the full mesh data. So go ahead, that would solve a lot of headaches!